src/coreclr/jit/lclvars.cpp

```diff
@@ -4437,7 +4470,13 @@ void Compiler::lvaComputeRefCounts(bool isRecompute, bool setSlotNumbers)
 
         // Set initial value for lvSingleDef for explicit and implicit
         // argument locals as they are "defined" on entry.
-        varDsc->lvSingleDef = varDsc->lvIsParam;
+        // However, if we are just recomputing the ref counts, retain the value
+        // that was set by past phases.
+        if (!isRecompute)
+        {
+            varDsc->lvSingleDef            = varDsc->lvIsParam;
```

Comment:
I don't quite understand this -- if we dead store something we can turn it from multiple def to single def. I suppose that won't happen for live into handler vars but it can for other vars.

Reply:
> I suppose that won't happen for live into handler vars but it can for other vars.

I am not 100% sure if that is the case. 

Currently, in the PR, I am marking `lvEhWriteThruCandidate` in one of the early phase of `lvaMarkLocalVars()`. I have to do it because that [information is used ](https://github.com/dotnet/runtime/blob/fc6823e4ae02a9bddbdf7b5b98e7c1a5c422970e/src/coreclr/jit/lclvars.cpp#L2626-L2629) during SSABuilder/Lowering to determine if a `varDsc` should be enregistered or not. The weird part of this logic for `lvaSetVarLiveInOutOfHandler()` is that during SSABuilder, we might mark a `varDsc->lvDoNotEnregister=1` but later during Lowering, if things change and we want to reset `lvDoNotEnregister` to enregister that EH variable, currently we don't do that. That should be fixed IMO. 

As per your suggestion, I should definitely take into account dead store, etc. and defer the setting of `lvEhWriteThruCandidate`  until I recompute the ref count during Lowering, but it would be too late and we might have already marked the `varDsc` to be "do not enregister". 

So perhaps, there might be some refactoring involved to accomplish things to happen in right order and I might want to do it in a separate PR. Other thing that I want to do in the follow-up PR is to support local fields, because [those are skipped ](https://github.com/dotnet/runtime/blob/fc6823e4ae02a9bddbdf7b5b98e7c1a5c422970e/src/coreclr/jit/lclvars.cpp#L4027-L4035) in this PR from getting marked as `lvEhWriteThruCandidate`. Perhaps, the logic of setting `lvEhWriteThruCandidate` should ideally be around [this code](https://github.com/dotnet/runtime/blob/fc6823e4ae02a9bddbdf7b5b98e7c1a5c422970e/src/coreclr/jit/lclvars.cpp#L4509).